# NumPy Basics
## Foundation Course MAS - HBRS
___Author : Iswariya Manivannan___

### Motivation
* __pandas__
* __scikit-learn__
* __tensorflow__
* __pytorch__

### Why np ?
* __ndarray__
* __array oriented arithmeic__
* __vectorized fast math functions__


### How to np ?

In [1]:
import numpy as np

 __ Write a code to__
 
* __sort "this"__
* __ find max of "this"__
* __ find the position of the max in "this"__

## this = [4,7,3,9,5] 


In [2]:
this = np.array([4,7,3,9,5])
np.max(this)

9

__Write a code to __
  *  __add "this" and "that"__
  *  __find max of this "this" and "that"__
  *  __add "5" with "that" __

## that = [10,20,30,40,50]


## Need for Speed

Compare numpy with random module and check the time of execution using time magic

In [6]:
%%time


CPU times: user 4 µs, sys: 2 µs, total: 6 µs
Wall time: 11.9 µs


## Getting help from numpy <br>
__What to do if shift+Tab doesnt work!!!__

In [7]:
np.info('sort')

     *** Found in numpy ***
 sort(a, axis=-1, kind='quicksort', order=None)

Return a sorted copy of an array.

Parameters
----------
a : array_like
    Array to be sorted.
axis : int or None, optional
    Axis along which to sort. If None, the array is flattened before
    sorting. The default is -1, which sorts along the last axis.
kind : {'quicksort', 'mergesort', 'heapsort', 'stable'}, optional
    Sorting algorithm. Default is 'quicksort'.
order : str or list of str, optional
    When `a` is an array with fields defined, this argument specifies
    which fields to compare first, second, etc.  A single field can
    be specified as a string, and not all fields need be specified,
    but unspecified fields will still be used, in the order in which
    they come up in the dtype, to break ties.

Returns
-------
sorted_array : ndarray
    Array of the same type and shape as `a`.

See Also
--------
ndarray.sort : Method to sort an array in-place.
argsort : Indirect sort.
lexsort : Indir

### The NumPy ndarray: A Multidimensional Array Object
* __from list__

In [8]:
list_array = np.array([1,2,3,4,5])


* __from list of list__      <br>Eg: matrix

In [9]:
lol_array = np.array([[1,2],[3,4],[5,6]])

* __Basic Details of an Array__  - shape, dimension, type

In [13]:
print " shape of array {}".format(lol_array.shape)
print " dimension of array {}".format(lol_array.ndim)
print " type of array {}".format(lol_array.dtype)

 shape of array (3, 2)
 dimension of array 2
 type of array int64


* __other ways of array creation__

Create an array of zeros of shape (2,2)

In [19]:
zeros = np.zeros((2,2), dtype=np.int64)
zeros.shape

(2, 3)

Create an array of ones of shape (3,4)

In [20]:
zeros = np.ones((3,4), dtype=np.int64)

Create an array of whole numbers ranging from 0 to 9

In [23]:
np.arange(0,10)

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

Create an array of even nos between 1 and 11

In [26]:
np.arange(2,12,2)

array([ 2,  4,  6,  8, 10])

Create an array of equally spaced 5 nos between 1 and 2. <br>
Hint: Stepsize is not known, so do not use np.arange() 

In [27]:
np.linspace(1,2,5)

array([1.  , 1.25, 1.5 , 1.75, 2.  ])

Create an empty array like another array

In [29]:
empty = np.empty_like(lol_array)
print empty


[[0 0]
 [0 0]
 [0 0]]


### Arithmetic with NumPy Arrays
__Vectorization__ <br>
__What does it mean? No need for loops__



Create an array 'this' which has the nos 2,10, 12

In [30]:
this = np.array([2,10,12])

Divide the elements of the array by 1/2

In [31]:
this/2

array([1, 5, 6])

Create an array 'that' which has the nos 4,5,6

In [33]:
that = np.array([4,5,6])

Compare the two arrays

In [34]:
this>that

array([False,  True,  True])

### Doctor Advise: Excercise is Good

* __Create two arrays A and B__

$A = \begin{bmatrix}1 & 2 & 3 \\4 & 5 & 6 \end{bmatrix}$ <br>

$B = \begin{bmatrix}3 & 7 & 2 \\0 & 2 & 4 \end{bmatrix}$ <br>

* Subtract the arrays from each other



In [36]:
A = np.array([[1,2,3],[4,5,6]])
B = np.array([[3,7,2],[0,2,4]])
A-B

array([[-2, -5,  1],
       [ 4,  3,  2]])

### Array Slicing
__sniper not periscope__

Create an array of nos from 0 to 10 and extract the first 5 nos.

In [42]:
arr = np.arange(0,11)
print arr
print arr.shape
arr[0:5]

[ 0  1  2  3  4  5  6  7  8  9 10]
(11,)


array([0, 1, 2, 3, 4])

Substitute the last 4 nos with zeros

In [47]:
arr[-4:] = 0
print arr

[0 1 2 3 4 5 6 0 0 0 0]


__Never loose the plot__

Extract every no in the even position

In [48]:
arr[::2]

array([0, 2, 4, 6, 0, 0])

__Reverse the array__

In [49]:
arr[::-1]

array([0, 0, 0, 0, 6, 5, 4, 3, 2, 1, 0])

__copy saves__

Create a copy which is not tied to the original array

In [60]:
arr2 = arr.copy()
arr2[0][1]=3
print arr, arr2

[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]] [[ 1  3  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


__Get used to 2D Array__

Create a 2D array of nos from 1 to 12 (shape of the array should be (4, 3))

In [51]:
arr = np.arange(1,13).reshape((4,3))
print arr


[[ 1  2  3]
 [ 4  5  6]
 [ 7  8  9]
 [10 11 12]]


### Arithmetic with NumPy Arrays
* 1D array __Broadcasting __ <br>
If two array differ in dimensions, then the shape of the array with fewer dimensions is padded with 1s on the left side.

Create an array of nos from 0 to 3

In [72]:
arr1 = np.array([0,1])

Add the two arrays and print the shape. What do you infer???

In [73]:
print(arr1.shape)
arr2 = np.arange(0,8).reshape((4,2))
add = arr2 + arr1
add.shape
print arr1.shape

(2,)
(2,)


* Multi-dimensional array __Broadcasting __ <br>
If two array differ in dimensions, then numpy tries to stretch the array wherever its shape is 1 to match the other array

Create an array of nos from 0 to 3 of shape (4, 1)

In [77]:
arr3 = np.arange(0,4).reshape((4,1))
print arr3.shape, arr3

(4, 1) [[0]
 [1]
 [2]
 [3]]


Multiply the two arrays and print the shape. What do you infer???

## Indexing numpy arrays

Extract the first element of the 2d array

In [81]:
sample = np.arange(0,20).reshape((4,5))
sample

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14],
       [15, 16, 17, 18, 19]])

Extract the first two rows of the array

In [80]:
sample[0:2]

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

Extract the first two columns of the array

In [90]:
sample[:,0:2]

array([[ 0,  1],
       [ 5,  6],
       [10, 11],
       [15, 16]])

Extract the elements in the first two rows and the first two columnst

In [91]:
sample[0:2,0:2]

array([[0, 1],
       [5, 6]])

### Doctor Advise: Excercise is Good

* __select the second row but only the first two columns__
* __select the third column but only the first two rows__
* __select all rows of the second column __

__[__<br\>
   __   [1, 0, 0],__ <br\>
__      [4, 0, 0],__ <br\>
__      [7, 8, 9]__<br\>
__]__<br\>

* __ Narrate the following slices:__<br\>
__arr [ : , : 2 ]__ <br\>
__arr [ 1, 1 : ]__ <br\>
__arr [ : ]__

### Bolean Indexing
Numpy arrays can also be indexed by boolean values!! <br>
__ my friends names in MOOC class and their assignment scores__

In [93]:
names = np.array(['ali','bob','chad','daisy'])
scores = np.array([
    [40,45,42],
    [10,15,7],
    [50,50,49],
    [50,2,50]
])
scores

array([[40, 45, 42],
       [10, 15,  7],
       [50, 50, 49],
       [50,  2, 50]])

__Daisy is my friend!! Make her scores 100!!__

Get the scores of daisy

In [94]:
scores[names=='daisy']

array([[50,  2, 50]])

Reset all scores of daisy to 100

* __Reshape__

Create an np array of nos from 0 to 31

In [98]:
arr1 = np.arange(0,32)

Reshape the array to a 4x8 matrix

In [101]:
arr2 = arr1.reshape((4,8))
print arr1

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31]


__hey, are you lazy? __  Provide the first index of the new np array and numpy takes care of it.

In [103]:
arr3 = arr1.reshape((4,-1))
print(arr3)

[[ 0  1  2  3  4  5  6  7]
 [ 8  9 10 11 12 13 14 15]
 [16 17 18 19 20 21 22 23]
 [24 25 26 27 28 29 30 31]]


Bring the array back to original size

In [105]:
arr4 = arr3.reshape(-1)
print arr4

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23
 24 25 26 27 28 29 30 31]


* __How to add a new dimension to an array ??__

In [109]:
arr = np.arange(0,32).reshape((4,-1))
print arr.ndim
arr = arr[:,:,np.newaxis]
print arr.shape

2
(4, 8, 1)


### Fancy Indexing - Revisited

**ATTENTION: The most wrong way of indexing and iterating through an numpy array!!!!**

Task: Create a numpy array of nos from 0 to 19 of shape 5x4. Print the rows of the array

**Instead use list of indices!! ** __pick the 1,3 rows__

__last two rows__

__guess the output__

In [ ]:
arr[[1,2,3],[3,2,1]]

__Still, we can make it!__

### Universal Functions: Fast Element-Wise Array Functions
These functions operate on an numpy array in an element wise manner and support broadcasting and type  casting <br>
Ref: https://docs.scipy.org/doc/numpy-1.14.0/reference/ufuncs.html

Create an array of nos from 0 to 10

__unary ufuncs__

Print the square and sqrt of the elements of the numpy array

Print the exponential of an array

Print the reciprocal of the elements of an array

__binary ufuncs__

Find the quotient and remainder

Check if two arrays are equal

Get the maximum of the two arrays

### Doctor Again! Excercise Again!

* date = ['01/01/2018','01/02/2018','01/03/2018','01/04/2018','01/05/2018'] <br\>
* stock_price = ['44','47.9','51','48.5','53']

#### quants suggested:
    * sell if price goes above 50 - Mark the output array with 'SELL'
    * buy if price goes less than 45 - Mark the output array with 'BUY'
    * hold if price is between 45 and 50 - Mark the output array with 'HOLD'
    
#### expected output: 
    * position : ['HOLD', 'BUY', 'SELL', 'BUY', 'SELL']

In [121]:
date = ['01/01/2018','01/02/2018','01/03/2018','01/04/2018','01/05/2018']
price = ['44','47.9','51','48.5','53']
date_arr = np.array(date)
price_arr = np.array(price, dtype=np.float)

output_arr = np.empty_like(date_arr)
output_arr[price_arr>50] = "SELL"
output_arr[price_arr<45] = "BUY"
output_arr[np.logical_and(price_arr>45,price_arr<50)] = "HOLD"
print output_arr


['BUY' 'HOLD' 'SELL' 'HOLD' 'SELL']


__when to hold?__

### Array-Oriented Programming with Arrays
__for-in-if __

Wherever the price is greater than 50 put 'SELL' otherwise replace with 'BUY'

**Not the right way to do it!!!**

**The right way!!!**

### ML
#### euclidean distance 
__$\sqrt{ (x1-x2)^2 + (y1-y2)^2 }$__

In [26]:
vector1 = np.array([1,2])
vector2 = np.array([3,2])


The Pythonic way!! Not to be used with numpy arrays

The Numpy way!!

Another easy way!!

### statistics

Generate an array of 4 random integers between 1 and 7

rand = np.random.randint(1,7,size=4)
print rand

Find the mean of the array

In [122]:
rand = np.random.randint(1,7,size=4)
print rand

[3 3 3 5]


Find the cummulative sum of elements

In [123]:
print np.mean(rand)
print np.max


3.5


Find the maximum element and its position.

In [ ]:
np.max

Generate an array of 10 random number from the normal distribution.

In [ ]:
np.random.randn

Find the number of elements greater than 0

In [124]:
a = np.arange(10)
print a[a>3]

[4 5 6 7 8 9]


### Sorting

Create a randomly permutated array of range 12 and shape 4,3

In [131]:
arrs = np.random.permutation(12).reshape((4,3))
print arrs


[[ 6  1 11]
 [ 0  2  5]
 [10  3  4]
 [ 7  8  9]]


Sort the elements of the array along columns

In [132]:
np.sort(arrs, axis=0)

array([[ 0,  1,  4],
       [ 6,  2,  5],
       [ 7,  3,  9],
       [10,  8, 11]])

Sort the elements of the array along rows

### set in Numpy

In [134]:
names = np.array(['Bob', 'Joe', 'Will', 'Bob', 'Will', 'Joe', 'Joe'])


Find the unique names in the array

In [135]:
np.unique(names)

array(['Bob', 'Joe', 'Will'], dtype='|S4')

### Linear Algebra

Create a numpy array of the matrix $A = \begin{bmatrix}1 & 2 & 3 \\4 & 5 & 6 \end{bmatrix}$

In [136]:
A = np.array([[1,2,3],[4,5,6]])

Create a numpy array of the matrix $B = \begin{bmatrix}2 & 2 \\1 & 1\end{bmatrix}$

In [138]:
B = np.array([[2,2],[1,1]])

Multiply the two matrices

Another way to multiply

In [141]:
C = np.dot(B,A)
print C

[[10 14 18]
 [ 5  7  9]]


Create a numpy array of the matrix $C = \begin{bmatrix}1 & 2 & 3 \\7 & 5 & 8 \end{bmatrix}$

In [146]:
C = np.array([[1,2,3],[7,5,8]])

Find the element-wise product (vector dot product or inner product) of A and C

Find the transpose of the matrix

Find the determinant of the matrix

Find the inverse of the matrix

Create an identity matrix of size 5x5

Extract the diagonal elements of matrix B

Create a diagonal matrix of the 1d array [3, 4, 5]

#### Distributions


Generate an array of 10 random nos sampled from the uniform distribution

Generate an array of size (2,2) sampled from the normal distribution

Generate an array of 10 random nos sampled from the uniform distribution with mean 2 and std 4

Generate an array of size (3,4) with nos sampled from the uniform distribution between 5 and 10

Generate an array of size (2,3) with nos sampled from the discrete uniform distribution between 1 and 8

### Further reading:
* Numpy docs
* Numpy 1.5 Beginners Guide
* Guide to Numpy http://web.mit.edu/dvp/Public/numpybook.pdf
* 100 Numpy exercises http://www.labri.fr/perso/nrougier/teaching/numpy.100/